[1.13] Dimension Registry

temp1011 opened the pull request:
Continued from https://github.com/ForestryMC/ForestryMC/pull/1939

### Problems
- If a mod wants to  blacklist dimensions by mod (e.g-no gen in Mystcraft ages or Rftools dimensions) they have no easy way to do this. Only blacklisting dimensions induvidually is realistically possible. a
At the moment, blacklisting by mod would involve something like `world.getWorldInfo.getWorldName()` but I suspect quite a few mods will use useless names (`miningWorld` or something) and this doesn't seem a good way to code things anyway.

- Furthermore, dimension IDs can't be guaranteed if a player adds/removes mods or mods creating dimensions on the fly. (I'm betting some mods use the ID `DimensionManager.getNextFreeDimId()` for their ID).

- Even if mods create their dimensions at startup it's possible they will try to use the same ID. While a correctly coded mod won't cause weird behaviour/crashes because of this it means the dimensions will be registered on different IDs to what might normally occur. This is then annoying because the pack maker/user/other mods may have to change their configs/mods.

- From what I can see, the current system makes it quite hard for other mods to keep track of which dimension is which because their main source of easily accessible information is the dimension `int` ID.

- Dimension IDs aren't very readable and don't provide much information in logs etc.

### Proposal
- Add a `BiHashMap<ResourceLocation, int>` to `DimensionManager` in Forge to act as a two way dictionary between the current dimension IDs (which I will call `dimensionIntID`s) and `ResourceLocation`s. **DONE**

- Add methods in `DimensionManager` to allow dimensions to be solely handled through Strings from the perspective of a mod and allow `ResourceLocation`s to be accessed. **DONE**

- `ResourceLocation`s will be constructed with arguments `modid` and `dimensionName` and `DimensionManager` will automatically assign `dimensionIntID`s to them. If a mod uses `dimensionIntID`s to register dimensions then they will be assigned `"noModid:dimension" + dimensionIntID` as a pair for the dictionary. **DONE**

- Deprecate using `dimensionIntIDs` in methods and instead suggest `ResourceLocation`s. **DONE**

- Add methods like `getDimensions(String modName)`, `isFromMod(String modName)` etc. which allow mods to blacklist their worldgen from others' dimensions. _Some added_

### Advantages
- Other mods can find dimensions even if not created at their usual dimension ID. 

- Any mod with their own worldgen (pretty much every mod) can control where it generates much more easily.

- Dimensions abstracted away from meaningless numbers.

- Easier to register dimensions (slightly, don't have to worry about which IDs are free).

### Disadvantages
- System where some use `int` and some `ResourceLocations` to register dimensions could be messy. However, mods should be able to keep using deprecated methods to deal with 'int' or this could be introduced as a breaking change.

- Mods will have to update their dimensions for this to have any use (making this a breaking change would solve this).

- Probably a bunch more I can't think of.

### To do
- Rewrite code of `DimensionManager` to use `ResourceLocation`s (at the moment using the new system will cause some performance lag. This means the dictionary will be the other way round by default and will be used to convert to IDs for Mojang's code.

- Turn 'DimensionManager` into an actual registry (would be a bit weird as some mods would want to register dimensions not at startup).

- Allow `"modid:name:type"` as an extra argument so mods can make clear what type of dimension they have (eg - void world, mining world, etc). _Would make things a bit more messy but doable_

- Add to IMC to let mods tell others not to do worldgen in their dimension.

- Test mod

**TLDR:** Some kind of dimension registry.

This probably needs to wait for 1.13 because Mojang do this themselves. Also, to  make the methods effective mods actually have to use the new system. It could be implemented in 1.12 (from what I can tell) but others would have to adopt it for the new features to have a big impact.

Sorry for making you read so much, it's partly to help me try to make sure I haven't missed something. There could be something obvious I've missed in which case let me know and I can try and fix it or close the PR. This code isn't amazing because I'm not an amazing coder but it should work. Any suggestions/feedback/improvements would be appreciated.

temp1011 committed:
Add string dictionary and add methods to use string

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/MinecraftForge/MinecraftForge?pullRequest=4666) <br/>All committers have signed the CLA.

liach commented on src/main/java/net/minecraftforge/common/DimensionManager.java:
```diff
@@ -103,25 +125,40 @@ public static void init()
+    
+    public static void registerDimension(String id, DimensionType type)
+    {
+    	if(id.indexOf(".")==-1)
```
use `contains` check instead

liach commented on src/main/java/net/minecraftforge/common/DimensionManager.java:
```diff
@@ -103,25 +125,40 @@ public static void init()
+    {
+    	if(id.indexOf(".")==-1)
+    	{
+    		throw new IllegalArgumentException(String.format("Failed to register dimension for stringID id %d, please use format modName.dimensionName", id));
```
We should use `modid:name` format, as in all resource locations

liach commented on src/main/java/net/minecraftforge/common/DimensionManager.java:
```diff
@@ -71,6 +73,7 @@ private Dimension(DimensionType type)
     private static Hashtable<Integer, WorldServer> worlds = new Hashtable<Integer, WorldServer>();
     private static boolean hasInit = false;
     private static Hashtable<Integer, Dimension> dimensions = new Hashtable<Integer, Dimension>();
+    private static HashBiMap<String, Integer> dimensionIDMap = HashBiMap.create();
```
```java
private static BiMap<String, Integer> dimensionIDMap = HashBiMap.create();
```
Just expose the interface type for variables

liach commented on src/main/java/net/minecraftforge/common/DimensionManager.java:
```diff
@@ -179,10 +249,50 @@ else if (leakCount % 5 == 0)
         return getIDs();
     }
+    
+    public static String[] getStringIDs(boolean check)
```
We can just do the int id check for leaks.

liach commented:
The whole thing looks big. It needs a detailed review and throughout testings.

temp1011 commented:
I agree it probably needs quite a bit of testing. Even if we are waiting for 1.13 I'll try and test this before so  we hopefully won't have to do it then unless code gets changed that this relies on.

temp1011 committed:
Quick improvements for lightning fast reviews

liach commented on src/main/java/net/minecraftforge/common/DimensionManager.java:
```diff
@@ -132,14 +132,14 @@ public static void init()
     
     public static void registerDimension(String id, DimensionType type)
     {
-    	if(id.indexOf(".")==-1)
```
Sorry, you should actually use `id.indexOf('.')` because checking a character is the fastest

temp1011 replied:
no worries :) but it's index of ":" right because we're using modid:name? Just checking for my sake.

liach replied:
Use single quotes

temp1011 committed:
fixed

temp1011 commented:
Is it worth some opinions of modders who use a lot of dimensions or may find this useful for their mods? Also, should I be aiming for using `String`s for the base to improve performance inside the class by not having to use the dictionary internally or should I keep the base workings as `int` so the code interfaces with Vanilla code without any conversions?
EDIT: Added deprecations to anything I wasn't sure was an internal method.

temp1011 committed:
These imports don't need to be here

temp1011 committed:
add deprecations

temp1011 committed:
fix String error in travis, add some new methods about dimensions and mods

liach commented on src/main/java/net/minecraftforge/common/DimensionManager.java:
```diff
@@ -140,9 +140,9 @@ public static void registerDimension(int id, DimensionType type)
     public static void registerDimension(String id, DimensionType type)
     {
-    	if(id.indexOf(":")==-1)
+    	if(id.indexOf(":")== -1 || id.indexOf(":")==0)
```
`(id.indexOf(':')== -1 || id.indexOf(':')==0)`

temp1011 committed:
single quotes

temp1011 commented:
Methods return arrays at the moment for multiple objects. Do we want to change this or does it give receiving mods more options? It's not really a big point but i'm somewhat interested as to the choice of array.

williewillus commented:
I think it's better to use resourcelocation for the ID's,  those are properly namespaced and are used as registry ID's for other registries.

temp1011 commented:
That could work well. then the stringid would also be the location of the file if I'm reading [this](https://mcforge.readthedocs.io/en/latest/concepts/resources/#resourcelocation) right  ([domain:path] instead of modid:name). I guess this also means we are going for a full implementation of a registry. i'll have a look at this.
EDIT: I'm a fool, I'll add this.

HenryLoenwind commented:
`"noModid:dimension" + dimensionIntID`---what about getCurrentModContainer?

temp1011 commented:
Could work
EDIT: But I can't find a function that does that. Would I be aiming to use an event?

temp1011 committed:
Now using ResourceLocations

temp1011 commented:
Sorry, my understanding of ResourceLocations at 2AM was not good and still might not be now. However, I think this should work.

temp1011 committed:
Returns Arrays of resourcelocation too

HenryLoenwind commented:
`Loader.instance().activeModContainer()` Have a look at `net.minecraftforge.fml.common.registry.IForgeRegistryEntry.Impl.setRegistryName(String)`, which basically does what you do here, too---setting a ResourceLocation as name on an object.
  

PS: Sorry I mixed up current/active in my original comment.

temp1011 commented:
No problem. Will the loader still return a useful value if a mod registers a dimension not during game load? I think I should just be creating a proper registry at this point anyway or at least extending `IForgeRegistryEntry.Impl`. Then again, I don't know if registries are really designed for adding entries while the game is running.

temp1011 committed:
Changing some internals to work off resourcelocations

HenryLoenwind commented:
Making a registry is actually really easy, at least of you don't need it to be persistant. [One of mine](https://github.com/SleepyTrousers/EnderIO/blob/master/enderio-base/src/main/java/crazypants/enderio/base/handler/darksteel/UpgradeRegistry.java)

temp1011 commented:
OK I could probably implement that. 

Why are mods doing this: https://github.com/zmaster587/AdvancedRocketry/blob/master/src/main/java/zmaster587/advancedRocketry/dimension/DimensionManager.java

temp1011 commented:
If I'm actually going to make dimensions a registry I can add to `ForgeRegistries` in the forge registry package right? Since this is something that should be made available to all mods.
@HenryLoenwind can I use your code to help with this?

temp1011 committed:
Big Change

temp1011 commented:
OK this breaks `DimensionManager` and probably a lot of other stuff too. I've made Dimensions into a `ForgeRegistry`(?). However, I don't have any events associated for this and I'm not sure if/how it's possible to modify this registry in the middle of the game (I did make sure to `allowModifications()` when creating the registry though). This is probably the point where my knowledge of how Forge works starts to get very thin (if it hasn't already). This is also a breaking change I'm pretty sure so we're definitely aiming for 1.13+ now. Do I need to add code so that mods can just pass a list of `Dimension`s at the right event or will Forge handle this? Should we also extend the `Dimension` class to give more options? (eg-override DimensionType in Vanilla)
EDIT: I think with one event during loading and then either using world events or some kind of dimension change event this can probably be solved.
EDIT:There already are dimension events I guess if they work.

temp1011 commented:
Hmm. I really not sure dimensions will work well like this. I'll see what happens but I may close this PR and open a fresh one to try and solve this in a better way. Maybe dimensions just shouldn't be fully implemented as a forge registry.

temp1011 committed:
Working on it

temp1011 committed:
Now breaks less

temp1011 commented:
Still broken but now I need to change patches so this gets more complicated
## Problems/Issues
- At the moment I don't know if the `int`s used to register dimensions in the registry at startup can be kept track of so they can be used later for mods that don't register at startup. I'm guessing this might be possible and if it is I probably won't have to change patches.
- If dimensions move completely to `ResourceLocation`s/`String`s I will need to re-patch: `WorldServer`,`WorldServerMulti`,`WorldProvider` and possibly `MinecraftServer` (can't quite remember). If I need to start patching things which weren't patched before (I don't think I should) it might get messy.
- Types in `DimensionManager` are a mess between `String` and `ResourceLocation`. I need to decide on one main one and stick to it.
- Active/Registered dimensions has likely been messed up somewhere. Need to check this.
- `Dimension` probably should still be a subclass of `DimensionManager` to stop mods accessing it and for tidiness.
- `REGISTRY` should probably be more public/accessible from `DimensionManager` (maybe `Dimension` as a `protected` subclass?)
- I can't work out where `getStaticDimensionIDs()` is called from. Is it still used?

Sorry to anyone reading this for the length and thanks for all the help. Feedback is always appreciated.

temp1011 committed:
Typos

temp1011 commented:
This is going to get pretty complex if I have to rewrite every single bit of code related to dimensions throughout Forge patches. Either I'm going to need to get which IDs registry has used to register dimensions at load or I'm going to have to figure it out myself. Essentially, I think it's going to be less of a hassle if the backend still runs in int (or at least the part that mods should be accessing I hope). Alternatively, I can rewrite pretty much all patches but checking this would be pretty bad. However, at least the types would be less messy then. Also, how do negative dimension IDs work? I assume it's just convention that people don't use them but `WorldProvider.shouldMapSpin()` may imply that they are used.

temp1011 commented:
I can't do 2+2=4, hopefully this will be sorted soon.

temp1011 committed:
Won't behave completely right but will build

temp1011 commented:
# Updated Situation
## The Good
- A registry for the class `Dimension` now exists and is part of  `ForgeRegistries`. However, it doesn't have a `RegistryEvent` yet (or the event won't point to the right place probably if it is automatic).
- `Dimension` is now separate from `DimensionManager` (but I'm considering putting it back inside `DimensionManager`). 
- Constructing a `Dimension` now requires a `String` ID. A `ResourceLocation` is then automatically assigned using `IForgeRegistryEntry` and an `int` ID is assigned using a `hashCode()` function that still maps Vanilla Dimensions to their correct places (I'm not anticipating many clashes). 
EDIT: Dammit there's a `bitset` for this.
- `DimensionManager` has methods for handling `Dimension`s based on `ResourceLocation`s so from this point of view mods shouldn't need to use any `int`s.
## The bad
- `Dimension` doesn't have a `RegistryEvent` yet (or the event won't point to the right place probably if it is automatic).
- Lots of methods using `int`s are still visible so mods might try and use these. Hopefully they won't if a nicer alternative exists.
- Something will probably be bugged in one of the backend classes somewhere. This will need testing/review to find.
- Similarly, some commands/methods may be returning `int`s to mods which are then annoying for them to handle.
- Don't know if reloading worlds will work at the moment for dimensions that registered while the game is running.
- Still more features to implement.
## The Ugly
- Mixing between `int`s and `ResourceLocation`s between different parts of code. If this is problematic in terms of what mods use I will change methods as needed.
- Probably some of my code in general 😢 .

temp1011 committed:
License header and unused imports/fields

temp1011 commented:
@mezz @liach sorry going to need some input on this one. As far as I can see the Forge Registries are designed to be not modified beyond startup. While it should be possible to use the default `ForgeRegistry` I don't know how well suited it is for the task. Would creating a custom `Registry` for this be a bad plan? Who would be the best person to talk to about this? Should I just create a new PR at this point to get rid of my ramblings. At the moment I seem to be merging about 3 different approaches which is not good. Thanks for any feedback and sorry for bothering you.
EDIT: Leaning towards getting rid of hashCode(), feels like collisions would be really annoying and can't really  handle it. However, using the standard Forge Registry I would then have to register some kind of dummy dimension in postinit() (I think) to get the free IDs easily.

liach commented:
I don't think you need to register dummy. Just take the overworld as the default one

temp1011 commented:
Does that mean I don't allow mods to register while the game is loading and instead get them to use the method in my `Dimension` class?
EDIT: You're right, I hadn't looked deep enough into the class. So I can guarantee the IDs are unique anyway. Thank you.

temp1011 committed:
I don't have to worry about int ID

temp1011 committed:
Beginning to sort out registry implementation

temp1011 committed:
Removing bitset save and load, adding a new function back in

I don't see why dimension ints need to be tracked now.

temp1011 commented:
## New Todo
- Stop registry tests failing (as they are now).
 - Allow "modid:name:type" as an extra argument so mods can make clear what type of dimension they have (eg - void world, mining world, etc). _Maybe will add_
- Add to IMC to let mods tell others not to do worldgen in their dimension.
- Test mod and test and review thoroughly (both client and server and both event and dynamic registrations) 
- Maybe this https://github.com/MinecraftForge/MinecraftForge/pull/4545

mezz commented:
I think it would be a good idea to open a new PR once this one is finished.
You are good at getting your ideas down for working this out, but a wall of text will turn away most people from reviewing a "finished" PR, so it would be good to start with a clean slate there.

temp1011 committed:
Part of vanilla works

temp1011 commented:
At this point minecraft will launch and the overworld and end will load fine, you can get back from the end fine, overworld saves data (as far as I can see). End may not because I was getting console spam about not finding a file related to it. Nether is broken (crash when going through a portal). This is probably because I had to assign it int ID 2 because the registry doesn't like negatives (maybe I should assign 255/256 and it won't crash). However hopefully this is just a case of changing a number somewhere. Haven't tested registering dimensions or anything on a server.

covers1624 commented on src/main/java/net/minecraftforge/common/DimensionManager.java:
```diff
@@ -128,12 +128,12 @@ public static WorldProvider getProvider(ResourceLocation dimID)
             for (ListIterator<World> li = allWorlds.listIterator(); li.hasNext(); )
             {
-                World w = li.next();
-                leakedWorlds.add(System.identityHashCode(w));
```
The entire point of leak detection is to use the identity hash of the world object, this breaks the entire leak detection system.

covers1624 commented on patches/minecraft/net/minecraft/init/Bootstrap.java.patch:
```diff
@@ -1,6 +1,14 @@
+ import net.minecraft.world.World;
+ import net.minecraft.world.biome.Biome;
+ import net.minecraft.world.storage.loot.LootTableList;
++import net.minecraftforge.common.Dimension;
```
Import in patch file.

covers1624 commented on src/main/java/net/minecraftforge/common/Dimension.java:
```diff
@@ -42,7 +42,7 @@ public Dimension(DimensionType type, String dimensionName)
     public static void registerDimensions()
     {
-    	registerDimension(-1, new Dimension(DimensionType.NETHER, "minecraft:nether"));
+    	registerDimension(2, new Dimension(DimensionType.NETHER, "minecraft:nether"));
```
Changing the id's of vanilla dimensions is probably an extremely bad idea. It will break compatibility with vanilla clients connecting to a forge server with no mods installed.

temp1011 replied:
True, but I'm not sure how to register this otherwise. I can shift all the ids up one and return one down I guess. Unless I'm missing something and I can register negative IDs? I think Dimensions are too different to be using the standard forge registry at this point.

temp1011 committed:
Updates

Game won't launch because registries hate negatives but I hope to fix this.

temp1011 committed:
Adapt registry for negative IDs

temp1011 commented:
I think I'm going to try to get registries which allow negatives. These will then be wrapped into some equivalent map to use. This could work as an extension/change to `ForgeRegistry` but at the moment I'm keeping it separate to not break too much. Probably some amount of adaptation needed in `GameData` too but that _should_ be fine.

As usual, any feedback is appreciated.

liach commented on patches/minecraft/net/minecraft/init/Bootstrap.java.patch:
```diff
@@ -10,15 +18,23 @@
+             PotionHelper.func_185207_a();
+             EntityList.func_151514_a();
+             Biome.func_185358_q();
++            Dimension.registerDimensions();
```
Move this to MinecraftForge class init method

liach commented on patches/minecraft/net/minecraft/server/MinecraftServer.java.patch:
```diff
@@ -15,7 +15,7 @@
 -    public long[][] field_71312_k;
 +    //public long[][] timeOfLastDimensionTick;
-+    public java.util.Hashtable<Integer, long[]> worldTickTimes = new java.util.Hashtable<Integer, long[]>();
++    public java.util.Hashtable<net.minecraft.util.ResourceLocation, long[]> worldTickTimes = new java.util.Hashtable<net.minecraft.util.ResourceLocation, long[]>();
```
Change Hashtable to Map

liach commented on patches/minecraft/net/minecraft/world/WorldProvider.java.patch:
```diff
@@ -82,14 +82,18 @@
 +    }
++    public int  getDimensionInt()
++    {
++    	return net.minecraftforge.common.Dimension.REGISTRY.func_82594_a(dimensionId).getDimIntID();
```
You can change so that the dimension int is kept in the dimension object instead of in the registry.

temp1011 replied:
I'm stupid... If I could like twice I would.

liach replied:
And you can keep a fastutil int2objectmap to map int dimensional numbers to dimension types

liach commented on src/main/java/net/minecraftforge/common/Dimension.java:
```diff
@@ -0,0 +1,96 @@
+    private final DimensionType type;
+    private final ResourceLocation dimID;
+    private int ticksWaited;
+    public static final RegistryNamespaced<ResourceLocation, Dimension> REGISTRY = GameData.getWrapper(Dimension.class);
```
Move this to `ForgeRegistries.DIMENSIONS` instead

liach commented on src/main/java/net/minecraftforge/common/Dimension.java:
```diff
@@ -0,0 +1,96 @@
+{
+    private final DimensionType type;
+    private final ResourceLocation dimID;
+    private int ticksWaited;
```
Add an int dim id field

liach commented on src/main/java/net/minecraftforge/common/Dimension.java:
```diff
@@ -0,0 +1,96 @@
+import net.minecraftforge.registries.GameData;
+import net.minecraftforge.registries.IForgeRegistryEntry;
+
+public class Dimension extends IForgeRegistryEntry.Impl<Dimension>
```
If you keep an dimID in your object, DO NOT extend Impl as that allows you to reset a registry name, causing extra work to be done

liach commented on src/main/java/net/minecraftforge/common/Dimension.java:
```diff
@@ -0,0 +1,96 @@
+    
+    public static void registerDimension(int intID, Dimension dimension)
+    {
+    	REGISTRY.register(intID, dimension.dimID, dimension);
```
This just shouldn't be used..

liach commented on src/main/java/net/minecraftforge/common/Dimension.java:
```diff
@@ -0,0 +1,96 @@
+
+    public static void registerDimension(Dimension dimension)
+    {
+    	REGISTRY.register(0, dimension.dimID, dimension);
```
Remove these registration methods. The registry events are always getting fired after preinit. You can move the registry event somewhere else if you wish to,

temp1011 replied:
Don't we have to account for mods that create dimensions in the middle of the game though eg-RFTools, Mystcraft etc? This might not be the right way though.

liach replied:
They can register dimensions in inits, but they can choose to load the level/world until the middle of the game.

liach replied:
(Much like what multiworld plugins do)

temp1011 replied:
I'll take a look

liach commented on src/main/java/net/minecraftforge/registries/ForgeRegistryExtended.java:
```diff
@@ -0,0 +1,984 @@
+import net.minecraftforge.fml.common.ModContainer;
+import net.minecraftforge.fml.relauncher.ReflectionHelper;
+
+public class ForgeRegistryExtended<V extends IForgeRegistryEntry<V>> implements IForgeRegistryInternal<V>, IForgeRegistryModifiable<V>
```
Why do you write this whole thing?

temp1011 replied:
Because I'm stupid and didn't work out how things work. It's mainly a copy of `ForgeRegistry`

liach replied:
It's fine

liach commented on src/main/java/net/minecraftforge/registries/GameData.java:
```diff
@@ -106,6 +108,8 @@
     private static final int MAX_RECIPE_ID = Integer.MAX_VALUE >> 5; // Varint CPacketRecipeInfo/SPacketRecipeBook
     private static final int MAX_PROFESSION_ID = 1024; //TODO: Is this serialized anywhere anymore?
+    private static final int MAX_DIMENSION_ID = Long.SIZE << 4;
+    private static final int MIN_DIMENSION_ID = -1;
```
Remove the min one

liach commented:
world and dimension are just different

temp1011 commented:
OK. I still want to implement IForgeRegistryEntry in my code so I can use the forge registry I think.

temp1011 committed:
Some changes so far

liach commented on src/main/java/net/minecraftforge/common/Dimension.java:
```diff
@@ -20,42 +20,29 @@
-    public static void registerDimension(int intID, Dimension dimension)
-    {
-    	REGISTRY.register(intID, dimension.dimID, dimension);
+    	if (getRegistryName() != null)
```
don't allow setting this. use dimid resource,location for registry name

temp1011 replied:
It's gone, just didn't save through to the commit.

temp1011 committed:
Hopefully better

temp1011 commented:
I don't know if it's better to let mods put their own `modid` in for the `ResourceLocation` or to get the `modid` automatically. At the moment I'm getting it myself, I'll change it if it somehow causes issues.

temp1011 committed:
Added map for int to dimension type

In dimension at the moment

temp1011 committed:
Added init, made constructor better (hopefully)

temp1011 committed:
Sorting out functions

Still have map of name to int in dimension to allow quick access from int->resourcelocation

temp1011 commented:
Stuff won't save correctly at the moment because the save folders use `ResourceLocations` since dimension IDs can change with adding/removing mods etc. Hopefully this just needs a couple of patches so the file manager can find the correct path. Maybe modifying file names is just a bad idea and will cause problems.
EDIT: From what I can see I'm not even creating any dim files inside my save. That's pretty bad

temp1011 committed:
Needed to stop clashes if my current plan works

temp1011 committed:
Saving now works, whether removing mods will is another question

temp1011 commented:
Save folders still use int IDs. This could make it tricky to reset dimensions (delete the file) for users but `/forge tps` let's people know which `int`s match to which dimensions. Still need to add events and other features.

Alpvax commented:
If the dimension is removed (i.e. the mod providing it is removed) any players in it should probably be moved to the overworld spawn (or their respawn point). If they keep their location but change dimension, they could be in what appears to be a completely new world (due to travel distance differences between dimensions).

Save folders should still use integers, at least for vanilla dimensions, due to vanilla compatibility.

temp1011 commented:
From the two mods I was able to quickly get set up I did get put back at spawn. Is that the mods code or does minecraft/forge handle this anyway?

liach commented on patches/minecraft/net/minecraft/world/WorldProvider.java.patch:
```diff
@@ -103,7 +103,7 @@
 +    public String getSaveFolder()
 +    {
-+        return (dimensionId.toString().equals("minecraft:overworld") ? null : "DIM" + dimensionId.toString());
++        return (dimensionId.toString().equals("minecraft:overworld") ? null : "DIM" + net.minecraftforge.fml.common.registry.ForgeRegistries.DIMENSIONS.getValue(dimensionId).getDimIntID());
```
You can just keep the dimension instance in worldprovider instead of looking up again from the registry

mcenderdragon commented:
what about `DimensionType` ? They already work alot like the ResourceLocations, but are weird enums

temp1011 commented:
It is an enum so I don't see how we can use it to identify dimensions since it can only take some set values and Forge doesn't know what/how many dimensions there are going to be.

temp1011 committed:
Dimension stored inside world provider

temp1011 committed:
Maybe int ID assigning

temp1011 commented:
This is my idea for getting `int` IDs to registered dimensions since that is what a lot of code still uses. This is probably quite ugly and there may be a better way to do this. I'm not sure if we are done registering by the init event or not which would be bad.

temp1011 committed:
Beginning of test mod

temp1011 committed:
Merge branch '1.12.x' into dimensionRegistry

temp1011 commented:
Starting on tests. At the moment my registry event doesn't seem to be working (the log output line is never printed) so that's pretty bad on my part. Hopefully I'll resolve it soon.

temp1011 committed:
Fix registering

temp1011 committed:
Sort out assigning ints, more test functionality

temp1011 committed:
Fix assigning ints..

Woops

temp1011 committed:
Working on fixing commands

Forge TPS is still broken but this should be on the way to fixing it. I think.

temp1011 committed:
Maybe just stick to ints...

temp1011 committed:
TPS Command fixed

temp1011 commented on src/main/java/net/minecraftforge/common/DimensionManager.java:
```diff
@@ -49,111 +47,82 @@
-    }
-
-    public static void init()
+    public static ResourceLocation[] getDimensions(DimensionType type)	//Could return Dimension[] since this contains name info
```
Should this be changed to return some kind of more general collection like https://github.com/MinecraftForge/MinecraftForge/pull/4545 since this will be a breaking change? Or should I follow that PR and just add another method to return a generic collection?

temp1011 commented on src/main/java/net/minecraftforge/common/DimensionManager.java:
```diff
@@ -49,111 +47,82 @@
     }
 
-    public static Integer[] getIDs(boolean check)
+    public static ResourceLocation[] getIDs(boolean check)
```
this

temp1011 commented on src/main/java/net/minecraftforge/common/DimensionManager.java:
```diff
@@ -445,4 +466,25 @@ else if (MinecraftServer.getServer() != null)
+    /* returns the dimensions registered to a mod based on its modid */
+    public static ResourceLocation[] getDimensionForMod(String modid)
+    {
+    	ResourceLocation[] ret = new ResourceLocation[ForgeRegistries.DIMENSIONS.getKeys().size()];
```
this and all the other examples

temp1011 reviewed (commented):
Should these be changed to return some kind of more general `Collection` like https://github.com/MinecraftForge/MinecraftForge/pull/4545 since this will be a breaking change? Or should I follow that PR and just add another method to return a generic `Collection`? I feel like there must be a better option than an `Array`

temp1011 committed:
Minor cleanup

liach commented:
Yes, you absolutely should.

temp1011 committed:
Change most return types to sets

Might help performance. Could go further and return collections. Might be some places where I've missed something.

temp1011 committed:
Reduce Code Duplication

temp1011 commented:
Not sure what the best way to handle loading dimensions on startup in `getStaticDimensionIDs()` is. Will probably have to save a list of dimensions somewhere or I could have something inside the `Dimension` object itself.
EDIT: I'm reading everything wrong, this should still be fine.

temp1011 committed:
woops

temp1011 committed:
Change method to use ForeRegistries better

temp1011 committed:
Merge branch '1.12.x' into dimensionRegistry

temp1011 commented:
I'm not sure when 1.13 will come out so I could add some stuff to 1.12. This would probably have to be opt in on the mod side. I think this has all the features I think it should at this point but maybe there are more suggestions.

mezz commented:
I was betting on 1.13 being released some time this month, but they have pushed it back with the combination of 1.13 and 1.14 so I am not really sure when it will be released either. I think it's probably better to just be patient with a big change like this, it is a lot easier to get it done the right way when you are not maintaining perfect backward compatibility.

temp1011 commented:
Not stale, just waiting for 1.13. If anyone uses dimensions a lot or has ideas for features/fixes feedback would be appreciated.

temp1011 commented:
Looks like 1.13 will change quite a bit with dimensions (there's a dimension class now it seems from what I've seen in SRG mappings). Hopefully that makes this require less patching and not more..

kashike commented:
@temp1011 `WorldProvider`s are now `Dimension`s - MCP classes have been renamed.

temp1011 commented:
Good change

temp1011 committed:
implement new persistent loading

temp1011 committed:
all the files

temp1011 committed:
Merge remote-tracking branch 'upstream/1.12.x' into dimensionRegistry

temp1011 committed:
fix merge problems

temp1011 committed:
compilation fixes

temp1011 committed:
fix tests

temp1011 commented:
Probably better to do this by cloning upstream, setting it up and then diffing the repos next time...

temp1011 committed:
fixes, cleanup, docs, tests

Alpvax commented on src/main/java/net/minecraftforge/common/Dimension.java:
```diff
@@ -29,26 +29,41 @@
-    public Dimension(DimensionType type, String dimensionName)
+
+	/**
+	 * Create a dimension using a dimension type alreadyd provided.
```
Typo: "alreadyd"

temp1011 committed:
typo

temp1011 commented:
I'm going to need to rename `Dimension` to avoid clashes with the new name for `WorldProvider`. Any suggestions?

temp1011 committed:
fix a potential bug

kashike commented on src/main/java/net/minecraftforge/common/Dimension.java:
```diff
@@ -0,0 +1,200 @@
+ * If the number of dimensions you need it variable it's probably best to register them at the start of the game and initialize them later
+ * @author temp1011 (needed?)
+ */
+public class Dimension implements IForgeRegistryEntry<Dimension>
```
This should have a different name; 1.13 uses this name, and I can see these being mixed up.

temp1011 replied:
Yeah I'm trying to think of a new one. `WorldInstance`? `DimensionInfo`? `DimensionProvider`? Not sure at the moment, I don't have anything particularly good.

mcenderdragon replied:
DimensionProvider sound not too bad.

kashike commented on src/main/java/net/minecraftforge/common/Dimension.java:
```diff
@@ -0,0 +1,200 @@
+	/**
+	 * default is -2, no dimension will be registered for this
+	 */
+	private int dimIntID = -2;
```
Your formatting is weird throughout this PR.

temp1011 replied:
Hopefully latest commit helps

temp1011 committed:
formatting

temp1011 committed:
more formatting, rename Dimension to DimensionProvider

temp1011 commented:
I've renamed 'Dimension' to 'DimensionProvider'. It may not be a good name or the final one but it won't clash.

temp1011 commented:
~~Refactoring has killed this, there are imports all over patch files now. Looking back, I'm not sure how long they were broken for. At least the diff size is back down now.~~

temp1011 committed:
fix imports

temp1011 committed:
re-add dimensionprovider

didn't stay watched through refactor

temp1011 committed:
remove author tag

kashike commented on patches/minecraft/net/minecraft/entity/Entity.java.patch:
```diff
@@ -80,6 +80,15 @@
+             p_189511_1_.func_74777_a("Air", (short)this.func_70086_ai());
+             p_189511_1_.func_74757_a("OnGround", this.field_70122_E);
+-            p_189511_1_.func_74768_a("Dimension", this.field_71093_bK);
++            p_189511_1_.func_74768_a("DimensionProvider", this.field_71093_bK);
```
No.

kashike commented on patches/minecraft/net/minecraft/entity/Entity.java.patch:
```diff
@@ -98,6 +107,18 @@
+             this.field_70122_E = p_70020_1_.func_74767_n("OnGround");
+ 
+-            if (p_70020_1_.func_74764_b("Dimension"))
++            if (p_70020_1_.func_74764_b("DimensionProvider"))
```
No.

kashike commented on patches/minecraft/net/minecraft/entity/Entity.java.patch:
```diff
@@ -98,6 +107,18 @@
++            if (p_70020_1_.func_74764_b("DimensionProvider"))
+             {
+-                this.field_71093_bK = p_70020_1_.func_74762_e("Dimension");
++                this.field_71093_bK = p_70020_1_.func_74762_e("DimensionProvider");
```
No.

kashike commented on patches/minecraft/net/minecraft/entity/Entity.java.patch:
```diff
@@ -146,6 +167,15 @@
+     {
+         NBTTagCompound nbttagcompound = p_180432_1_.func_189511_e(new NBTTagCompound());
+-        nbttagcompound.func_82580_o("Dimension");
++        nbttagcompound.func_82580_o("DimensionProvider");
```
No.

temp1011 committed:
Clean NBT tag names

temp1011 commented:
Yes?

jamierocks commented:
There are still plenty of issues from your Dimension -> DimensionProvider refactor.

temp1011 commented:
This probably is still broken/incomplete but this is all I have time for now. I'll squash once I'm done.

jamierocks commented on patches/minecraft/net/minecraft/world/WorldProvider.java.patch:
```diff
@@ -65,30 +65,34 @@
-+        return this.dimensionId;
++        return dimensionProvider.getID();
++    }
++    public int getDimensionProviderInt()
```
This is a poor name, perhaps `getDimensionId()`?

temp1011 replied:
`WorldProvider` is going to be renamed to `Dimension` in 1.13 which might be a problem. Also, I would like to make the `int` and `ResourceLocation` ids quite separate as hopefully the `int` should only be used internally. I agree that the name isn't great at the moment.

jamierocks replied:
My point is more that it isn't the ID of the provider, but the underlying dimension.

temp1011 replied:
Yes, so changing it to `getDimensionID()` would not make a lot of sense when `WorldProvider` is renamed? This will definitely only get merged in 1.13 so I'm basing names around the MCP mappings there.

temp1011 commented:
I've broken history so need to reset.

temp1011 committed:
refactor

temp1011 commented:
Now that 1.13.2 uses a resourcelocation to register dimensions this should no longer be necessary.
